File: node.js/RPSGame.js
'use strict';

//let pserv = require('./promise-server');
let net = require('net');
let pcli = require('./promise-client');
let Hash = require('./hash');
let udpe = require('./RPSUDPEmitter');

class RPS {
	constructor(v) {
		this.v = v;
	}

	beats(other) {
		if (this.v == RPS.Lose.v)
			return false;
		switch (this.v) {
		case 0:
			return other.v == 2;
		case 1:
			return other.v == 0;
		case 2:
			return other.v == 1;
		default:
			return false;
		}
	}

	toString() {
		if (this.v < 0 || this.v > 2)
			return "Lose";
		return ["Rock", "Paper", "Scissors"][this.v];
	}
}

RPS.Lose = new RPS(255);
RPS.Rock = new RPS(0);
RPS.Paper = new RPS(1);
RPS.Scissors = new RPS(2);

let delay = n => {
	return new Promise((res, rej) => {
		setTimeout(() => res(), n)
	});
}

class RPSGame {
	constructor(host) {
		if (host !== undefined)
			this.host = host;
		this.players = [];
		this.pchoices = [];
		this.ppoints = [];
		this.mypoints = 0;
	}

	async host(n) {
		let uc = new udpe();
		let completePlayers = [];
		let hc = net.createServer(async sock => {
			if (n <= 0) {
				sock.end();
				return;
			}
			let newp = new pcli(sock);
			newp.gport = await newp.read(2);
			await newp.write(Buffer.from([0x43, completePlayers.length]));
			let tasks = completePlayers.map(async p => {
				let addr = p.client.remoteAddress.split('.').map(i => parseInt(i));
				let port = p.gport;
				await newp.write([...addr, ...port]);
			});
			await Promise.all(tasks);
			completePlayers.push(newp);
			n--;
		});
		hc.listen({host: '0.0.0.0', port: 7853}, () => {});
		while (n > 0) {
			await uc.write([0x44, 0x49, 0x53, n]);
			await delay(1000);
		}
		await Promise.all(completePlayers.map(async p => {
			p.write(Buffer.from([ 0x4F, 0x4B, completePlayers.length % 255 ]));
			p.close();
		}));
		uc.close();
		hc.close();
	}

	choose() {
		return new Promise((res, rej) => {
			let selected = 1;
			let choices = [ "💎", "📜", "✂️" ];
			let values = [ RPS.Rock, RPS.Paper, RPS.Scissors ];
			let k;

			let printSelection = () => {
				process.stdout.write("\x1b[2K\r");
				for (let i = 0; i < 3; ++i) {
					if (i == selected) {
						process.stdout.write("\x1b[7m");
					}
					process.stdout.write(choices[i] + " ");
					if (i == selected) {
						process.stdout.write("\x1b[0m");
					}
					process.stdout.write(" ");
				}
				process.stdout.write("arrows to select, space to confirm");
			}

			process.stdin.setRawMode(true);
			process.stdin.setEncoding('utf8');

			process.stdout.write('\x1b[?25l');
			printSelection();

			let handleInput = k => {
				if (k == "\u001b[D")
					selected = (selected - 1);
				else if (k == "\u001b[C")
					selected = (selected + 1) % 3;
				if (selected < 0)
					selected = 2;
				printSelection();
				if (k == ' ') {
					process.stdout.write('\x1b[?25h');
					process.stdin.setRawMode(false);
					process.stdin.removeAllListeners('data');
					res(values[selected]);
				}
			}

			process.stdin.on('data', handleInput);
		});
	}

	random(n) {
		return [...Array(n).keys()].map(() => ~~(Math.random() * 255));
	}

	async doTurn() {
		let choice = await this.choose();
		console.log(`\nI chose ${choice}`);
		let packet = this.random(31);
		packet.unshift(choice.v);
		let hash = Hash.hashData(packet);
		hash.unshift(0x4B);

		let choicestasks = this.players.map(async cli => {
			await cli.write(Buffer.from(hash));
			let oh = await cli.read(33);
			// Point de synchro
			if (oh[0] != 0x4B) {
				return RPS.Lose;
			}
			let otherHash = oh.slice(1);
			oh = [0x56, choice.v];
			oh.push(...packet.slice(1));
			await cli.write(Buffer.from(oh));
			oh = await cli.read(33);
			if (oh[0] != 0x56) {
				return RPS.Lose;
			}
			let otherChoice = new RPS(oh[1]);
			console.log("An other player supposedly chose " + otherChoice);
			let otherSecret = oh.slice(2);
			otherSecret.unshift(otherChoice.v);
			let longHash = Hash.hashData(otherSecret);
			if (otherHash.every((v, i) => v === longHash[i]))
				return otherChoice;
			return RPS.Lose;
		});

		choicestasks = await Promise.all(choicestasks);
		this.pchoices = choicestasks;
		let sum = (b, c) => b + c;
		this.ppoints = this.pchoices.map(c => {
			if (c == RPS.Lose)
				return -1;
			let wins = this.pchoices.map(d => c.beats(d)).reduce(sum);
			let losses = this.pchoices.map(d => d.beats(c)).reduce(sum);
			wins += c.beats(choice);
			losses += choice.beats(c);
			return wins - losses;
		});
		let mwins = this.pchoices.map(d => choice.beats(d)).reduce(sum);
		let mlosses = this.pchoices.map(d => d.beats(choice)).reduce(sum);
		this.mypoints = mwins - mlosses;
	}

	async play() {
		while (this.players.length) {
			console.log(`There's ${this.players.length} other players left!`);
			this.ppoints = [];
			await this.doTurn();
			this.players.map((p, idx) => [p, idx])
				.filter(n => this.ppoints[n[1]] < 0)
				.forEach(n => {
					console.log(`Removing a player because of sub ${this.ppoints[n[1]]} points`);
					this.ppoints.splice(n[1], 1);
					this.players[n[1]].close();
					this.players.splice(n[1], 1);
				})
			if (this.mypoints < 0) {
				console.log("I lost to this");
				break;
			}
			var endtasks = this.players.map(async p => {
				let data = [0x43, 0x4F, 0x4E];
				await p.write(data);
				let res = await p.read(3);
				return res.every((v, i) => v === data[i]);
			});
			// TODO: timeout here
			await Promise.all(endtasks);
			let results = endtasks.map(t => {
				if (!t)
					Console.WriteLine("Other player violated protocol");
				return t;
			});
		}
		if (this.players.length == 0)
			console.log(`You won with ${this.mypoints} points!`);
		else
			console.log(`You lost with ${this.mypoints} points!`);
		this.players.forEach(p => p.close());
		this.players = [];
	}

	async prepare() {
		let playersips = [];
		let hostcon = new pcli(this.host.address, this.host.port);
		let serv = net.createServer(sock => this.players.push(new pcli(sock)));
		let port = await new Promise(res => {
			serv.listen({port: 0}, async () => {
				let nport = serv.address().port
				nport = [nport & 0xFF, nport >> 8 & 0xFF];
				res(nport);
			});
		});
		await hostcon.write(port);
		let head = await hostcon.read(2);
		if (head[0] != 0x43)
			throw new Error("Protocol error");
		let i = 0;
		while (i < head[1]) {
			let np = await hostcon.read(6);
			let pip = {
				address: `${np[0]}.${np[1]}.${np[2]}.${np[3]}`,
				port: np[5] << 8 | np[4]
			};
			console.log(JSON.stringify(pip));
			playersips.push(pip);
			++i;
		}
		this.players = playersips.map(ip => new pcli(ip.address, ip.port));
		while (true) {
			let beg = await hostcon.read(1);
			if (beg[0] == 0x4F) {
				let res = await hostcon.read(2);
				console.log(res);
				if (res[0] == 0x4B) {
					console.log(`Starting game with ${res[1]} other players`);
					hostcon.close();
					serv.close();
					break;
				}
			} else {
				console.log("Unknown request", beg);
			}
		}
	}
}

module.exports = RPSGame;

File: C#.NetOnMono/rps_nett/RPSGame.cs
using System;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace rps_nett {
    public class RPSGame {
        public List<TcpClient> players = new List<TcpClient>();
        List<RPS> playersChoices = new List<RPS>();
        List<int> playersPoints = new List<int>();
        int myPoints;
        private IPEndPoint host;

        public RPSGame(IPEndPoint host) {
            this.host = host;
        }

        public RPSGame() {
        }

        public async Task Host(int n) {
            var udpEp = new IPEndPoint(IPAddress.Broadcast, 7854);
            var uc = new UdpClient(udpEp) {
                EnableBroadcast = true
            };
            uc.Connect(udpEp);
            var hc = new TcpListener(new IPEndPoint(IPAddress.Any, 7853));
            hc.Start();
            var acceptTask = hc.AcceptTcpClientAsync();
            var completePlayers = new List<(TcpClient, byte[])>();

            while (n > 0) {
                await uc.SendAsync(new byte[] { 0x44, 0x49, 0x53, (byte)n }, 4);
                var finished = await Task.WhenAny(acceptTask, Task.Delay(1000));
                if (finished == acceptTask) {
                    var newPlayer = acceptTask.Result;
                    var newPlayerCon = new TcpListener(IPAddress.Any, 0);
                    newPlayerCon.Start();
                    var npportb = new byte[2];
                    await newPlayer.GetStream().ReadAsync(npportb, 0, 2);
                    var tuples = completePlayers.ConvertAll(p => {
                        var stream = p.Item1.GetStream();
                        var tuple = new byte[6];
                        var ip = (IPEndPoint)p.Item1.Client.RemoteEndPoint;
                        ip.Address.GetAddressBytes().CopyTo(tuple, 0);
                        p.Item2.CopyTo(tuple, 4);
                        return tuple;
                    });
                    await newPlayer.GetStream().WriteAsync(new byte[] { 0x43, (byte)(tuples.Count) }, 0, 2);
                    tuples.ForEach(t => newPlayer.GetStream().WriteAsync(t, 0, 6));
                    completePlayers.Add((newPlayer, npportb)); // ???
                    --n;
                    acceptTask = hc.AcceptTcpClientAsync();
                }
            }
            uc.Close();
            hc.Stop();
            var endMessage = new byte[3] { 0x4F, 0x4B, (byte)(completePlayers.Count) };
            Task.WaitAll(completePlayers.ConvertAll(async p => {
                await p.Item1.GetStream().WriteAsync(endMessage, 0, 3);
                p.Item1.Close();
            }).ToArray());
        }

        public async Task Prepare() {
            List<IPEndPoint> playersips = new List<IPEndPoint>();
            var listener = new TcpListener(IPAddress.Any, 0);
            listener.Start();

            var hostcon = new TcpClient();
            hostcon.Connect(host);
            var stream = hostcon.GetStream();
            var pport = ((IPEndPoint)listener.Server.LocalEndPoint).Port;
            await stream.WriteAsync(BitConverter.GetBytes(pport), 0, 2);

            byte[] detail = new byte[3];
            await stream.ReadAsync(detail, 0, 2);
            if (detail[0] != 0x43)
                throw new Exception("Badly formatted packet");
            int i = 0;
            while (i < detail[1]) {
                byte[] newPlayer = new byte[6];
                stream.Read(newPlayer, 0, 6);
                int addr = BitConverter.ToInt32(newPlayer, 0);
                ushort port = BitConverter.ToUInt16(newPlayer, 4);
                playersips.Add(new IPEndPoint(addr, port));
                ++i;
            }
            players = playersips.ConvertAll(ip => {
                var r = new TcpClient();
                Console.WriteLine("Connecting to another player...");
                r.Connect(ip);
                return r;
            });
            detail[0] = 0;
            while (true) {
                Console.WriteLine("Waiting for beginning");
                stream.Read(detail, 0, 3);
                if (detail[0] == 0x4F) {
                    if (detail[1] == 0x4B) {
                        while (players.Count() != detail[2] - 1) {
                            players.Add(await listener.AcceptTcpClientAsync());
                        }
                        listener.Stop();
                        Console.WriteLine("Starting game with {0:d} other players", detail[2] - 1);
                        break;
                    }
                }
            }
        }

        RPS Choose() {
            ConsoleKey k;
            int selected = 1;
            string[] choices = { "💎", "📜", "✂️" };
            Console.CursorVisible = false;
            do {
                Console.Write("\x1b[2K\r");
                for (int i = 0; i < 3; ++i) {
                    // bug graphique où noir est différent de la couleur de base
                    // peut etre du au fait que mon terminal supporte 256 couleurs
                    if (i == selected) {
                        var tmp = Console.BackgroundColor;
                        Console.BackgroundColor = Console.ForegroundColor;
                        Console.ForegroundColor = tmp;
                    }
                    Console.Write(choices[i] + " ");
                    if (i == selected) {
                        var tmp = Console.BackgroundColor;
                        Console.BackgroundColor = Console.ForegroundColor;
                        Console.ForegroundColor = tmp;
                    }
                    Console.Write(" ");
                }
                Console.Write("arrows to select, space to confirm");
                k = Console.ReadKey(false).Key;
                if (k == ConsoleKey.LeftArrow)
                    selected = (selected - 1);
                else if (k == ConsoleKey.RightArrow)
                    selected = (selected + 1) % 3;
                if (selected < 0)
                    selected = 2;
            } while (k != ConsoleKey.Spacebar);
            Console.CursorVisible = true;
            byte choice = (byte)(selected);
            Console.WriteLine("\nI chose " + ((RPS)choice));
            return choice;
        }

        void DoTurn() {
            var choice = Choose();
            byte[] packet = new byte[32];
            byte[] secret = new byte[31];
            byte[] full = new byte[33];
            new Random().NextBytes(secret);
            packet[0] = choice;
            Buffer.BlockCopy(secret, 0, packet, 1, 31);
            var hash = MainClass.HashData(packet);
            full[0] = 0x4B;
            Buffer.BlockCopy(hash, 0, full, 1, 32);
            Console.WriteLine("Using secret " + MainClass.BytesToString(secret));
            Console.WriteLine("Sending packet " + MainClass.BytesToString(full));

            var choicestasks = players.ConvertAll(async cli => {
                byte[] oh = new byte[33];
                byte[] otherHash = new byte[32];
                byte[] otherSecret = new byte[32];
                var cs = cli.GetStream();
                await cs.WriteAsync(full, 0, 33);
                await cs.ReadAsync(oh, 0, 33);
                // Point de synchro
                if (oh[0] != 0x4B) {
                    Console.WriteLine("Malformed response...");
                    return RPS.Lose;
                }
                Console.WriteLine("Received packet " + MainClass.BytesToString(oh));
                Buffer.BlockCopy(oh, 1, otherHash, 0, 32);
                oh[0] = 0x56;
                oh[1] = choice;
                Buffer.BlockCopy(secret, 0, oh, 2, 31);
                Console.WriteLine("Sending verification packet " + MainClass.BytesToString(oh));
                await cs.WriteAsync(oh, 0, 33);
                await cs.ReadAsync(oh, 0, 33);
                Console.WriteLine("Received verification packet " + MainClass.BytesToString(oh));
                if (oh[0] != 0x56) {
                    Console.WriteLine("Malformed response...");
                    return RPS.Lose;
                }
                RPS otherChoice = oh[1];
                Console.WriteLine("Other player supposedly chose " + otherChoice);
                Buffer.BlockCopy(oh, 2, otherSecret, 1, 31);
                otherSecret[0] = otherChoice;
                var longHash = MainClass.HashData(otherSecret);
                if (otherHash.SequenceEqual(longHash))
                    return otherChoice;
                Console.WriteLine("Bad hash detected");
                return RPS.Lose;
            });

            Task.WaitAll(choicestasks.ToArray(), 500000);
            playersChoices = choicestasks.ConvertAll(t => {
                if (!t.IsCompleted)
                    Console.WriteLine("Player took too long to answer");
                return t.IsCompleted ? t.Result : RPS.Lose;
            });
            playersPoints = playersChoices.ConvertAll(c => {
                if (c == RPS.Lose)
                    return -1;
                var wins = playersChoices.ConvertAll(d => c.Beats(d) ? 1 : 0).Sum();
                var losses = playersChoices.ConvertAll(d => d.Beats(c) ? 1 : 0).Sum();
                wins += c.Beats(choice) ? 1 : 0;
                losses += choice.Beats(c) ? 1 : 0;
                return wins - losses;
            });
            var mwins = playersChoices.ConvertAll(d => choice.Beats(d) ? 1 : 0).Sum();
            var mlosses = playersChoices.ConvertAll(d => d.Beats(choice) ? 1 : 0).Sum();
            myPoints = mwins - mlosses;
        }

        public void Play() {
            while (players.Count != 0) {
                Console.WriteLine("There's {0:d} other players left!", players.Count);
                playersPoints = new List<int>();
                DoTurn();

                var zipped = players.Zip(playersPoints, (TcpClient other, int points) => new { other, points });
                var removed = zipped.Where(z => z.points < 0);
                var alive = zipped.Where(z => z.points >= 0);
                players = alive.Select(z => z.other).ToList();
                playersPoints = alive.Select(z => z.points).ToList();

                foreach (var z in removed) {
                    z.other.Close();
                }
                if (myPoints < 0) {
                    Console.WriteLine("I lost to this");
                    break;
                }
            }
            if (players.Count == 0)
                Console.WriteLine("You won with {0:d} points!", myPoints);
            else
                Console.WriteLine("You lost with {0:d} points!", myPoints);
            Console.WriteLine("Ended with {0:d} players left!", players.Count);
        }

        private class RPS {
            byte b;

            public static RPS Lose = new RPS { b = 255 };
            public static RPS Rock = new RPS { b = 0 };
            public static RPS Paper = new RPS { b = 1 };
            public static RPS Scissors = new RPS { b = 2 };

            private RPS() { }

            public static implicit operator byte(RPS r) {
                return r.b;
            }

            public override string ToString() {
                if (b >= 3)
                    return "Lose";
                return new string[] { "Rock", "Paper", "Scissors" }[b];
            }

            public bool Beats(RPS o) {
                if (this == Lose)
                    return false;
                switch (b) {
                case 0:
                    return o == Scissors;
                case 1:
                    return o == Rock;
                case 2:
                    return o == Paper;
                default:
                    return false;
                }
            }

            public bool IsBeatenBy(RPS o) {
                if (this == Lose)
                    return true;
                return this != o && !Beats(o);
            }

            public static implicit operator RPS(byte r) {
                switch (r) {
                case 0:
                    return Rock;
                case 1:
                    return Paper;
                case 2:
                    return Scissors;
                default:
                    return Lose;
                }
            }
        }
    }
}